executor/merge_join.go

```diff
@@ -215,196 +269,114 @@ func (t *mergeJoinInnerTable) hasNullInJoinKey(row chunk.Row) bool {
 func (e *MergeJoinExec) Open(ctx context.Context) error {
 	if err := e.baseExecutor.Open(ctx); err != nil {
 		return err
 	}
 
-	e.prepared = false
 	e.memTracker = memory.NewTracker(e.id, e.ctx.GetSessionVars().MemQuotaMergeJoin)
 	e.memTracker.AttachTo(e.ctx.GetSessionVars().StmtCtx.MemTracker)
+	e.diskTracker = disk.NewTracker(e.id, -1)
+	e.diskTracker.AttachTo(e.ctx.GetSessionVars().StmtCtx.DiskTracker)
 
-	e.childrenResults = make([]*chunk.Chunk, 0, len(e.children))
-	for _, child := range e.children {
-		e.childrenResults = append(e.childrenResults, newFirstChunk(child))
-	}
-
-	e.innerTable.memTracker = memory.NewTracker(innerTableLabel, -1)
-	e.innerTable.memTracker.AttachTo(e.memTracker)
-
-	return nil
-}
-
-func compareChunkRow(cmpFuncs []chunk.CompareFunc, lhsRow, rhsRow chunk.Row, lhsKey, rhsKey []*expression.Column) int {
-	for i := range lhsKey {
-		cmp := cmpFuncs[i](lhsRow, lhsKey[i].Index, rhsRow, rhsKey[i].Index)
-		if cmp != 0 {
-			return cmp
-		}
-	}
-	return 0
-}
-
-func (e *MergeJoinExec) prepare(ctx context.Context, requiredRows int) error {
-	err := e.innerTable.init(ctx, e.ctx, e.childrenResults[e.outerIdx^1])
-	if err != nil {
-		return err
-	}
-
-	err = e.fetchNextInnerRows()
-	if err != nil {
-		return err
-	}
-
-	// init outer table.
-	e.outerTable.chk = e.childrenResults[e.outerIdx]
-	e.outerTable.iter = chunk.NewIterator4Chunk(e.outerTable.chk)
-	e.outerTable.selected = make([]bool, 0, e.maxChunkSize)
-
-	err = e.fetchNextOuterRows(ctx, requiredRows)
-	if err != nil {
-		return err
-	}
-
-	e.prepared = true
+	e.innerTable.init(e)
+	e.outerTable.init(e)
 	return nil
 }
 
 // Next implements the Executor Next interface.
-func (e *MergeJoinExec) Next(ctx context.Context, req *chunk.Chunk) error {
+func (e *MergeJoinExec) Next(ctx context.Context, req *chunk.Chunk) (err error) {
 	req.Reset()
-	if !e.prepared {
-		if err := e.prepare(ctx, req.RequiredRows()); err != nil {
-			return err
-		}
-	}
 
+	innerIter := e.innerTable.groupRowsIter
+	outerIter := e.outerTable.groupRowsIter
 	for !req.IsFull() {
-		hasMore, err := e.joinToChunk(ctx, req)
-		if err != nil || !hasMore {
-			return err
+		if innerIter.Current() == innerIter.End() {
+			if err := e.innerTable.fetchNextInnerGroup(ctx, e); err != nil {
+				return err
+			}
+			innerIter = e.innerTable.groupRowsIter
 		}
-	}
-	return nil
-}
-
-func (e *MergeJoinExec) joinToChunk(ctx context.Context, chk *chunk.Chunk) (hasMore bool, err error) {
-	for {
-		if e.outerTable.row == e.outerTable.iter.End() {
-			err = e.fetchNextOuterRows(ctx, chk.RequiredRows()-chk.NumRows())
-			if err != nil || e.outerTable.chk.NumRows() == 0 {
-				return false, err
+		if outerIter.Current() == outerIter.End() {
+			if err := e.outerTable.fetchNextOuterGroup(ctx, e, req.RequiredRows()-req.NumRows()); err != nil {
+				return err
+			}
+			outerIter = e.outerTable.groupRowsIter
+			// no more data to execute
+			if outerIter.Current() == outerIter.End() {
```

Comment:
How about replacing this condition with `e.outerTable.executed`, which seems easier to understand?

Reply:
fixed